File: src/main/java/br/com/pnp/compiler/model/construct/statement/IfStatement.java
package br.com.pnp.compiler.model.construct.statement;

import br.com.pnp.compiler.model.construct.Block;
import br.com.pnp.compiler.model.construct.type.Type;
import br.com.pnp.compiler.model.construct.type.primitives.PrimitiveType;
import br.com.pnp.compiler.model.expression.Expression;

public class IfStatement extends Statement {
    private Block block;
    private IfStatement elseStatement = null;
    private Expression condition;

    public IfStatement(Expression condition) {
        this.block = new StatementBlock();
        this.condition = condition;
    }

    private IfStatement(IfStatement ifBranch) {
        this.block = new StatementBlock();
    }

    public Block startElseBranch(Expression condition) {
        elseStatement = new IfStatement(condition);
        return elseStatement.block;
    }

    public Block startElseBranch() {
        elseStatement = new IfStatement(this);
        return elseStatement.block;
    }

    public boolean hasElseBranch() {
        return elseStatement != null;
    }

    @Override
    public Type getType() {
        return PrimitiveType.Booleano;
    }

    public Block getIfBlock() {
        return block;
    }

    public Block getElseBlock() {
        if (elseStatement != null) {
            return elseStatement.block;
        }
        return null;
    }

    public IfStatement getElseStatement() {
        return elseStatement;
    }

    public Expression getCondition() {
        return condition;
    }
}

File: src/main/kotlin/br/com/pnp/model/construct/statement/IfStatement.kt
package br.com.pnp.model.construct.statement

import br.com.pnp.model.construct.Block
import br.com.pnp.model.construct.Variable
import br.com.pnp.model.construct.type.primitive.PrimitiveType
import br.com.pnp.model.expression.Expression

class IfStatement(condition: Expression): Statement {
    val ifBlock: Block = StatementBlock()
    val elseBlock: Block? get() = elseStatement?.ifBlock
    val hasElseBranch: Boolean get() = elseStatement != null

    private var elseStatement: IfStatement? = null
    val elseBranch: IfStatement? get() = elseStatement

    var condition: Expression = Variable.literalBoolean(true)
        set(value) {
            if (!PrimitiveType.boolean.isTypeOf(value.type)) {
                throw IllegalArgumentException(
                        """
                        Type mismatch in ${this.javaClass.simpleName} constructor:
                        Required: ${PrimitiveType.boolean}
                        Found: ${value.type}
                    """.trimIndent())
            }
            field = value
        }

    init {
        this.condition = condition
    }

    private constructor(): this(Variable.literalBoolean(true))

    fun startElseStatement(): Block {
        val newStatement = IfStatement()
        val ifStatement = lastElseStatement()

        ifStatement.elseStatement = newStatement
        return newStatement.ifBlock
    }

    fun startElseStatement(condition: Expression): Block {
        val newStatement = IfStatement(condition)
        val ifStatement = lastElseStatement()

        ifStatement.elseStatement = newStatement
        return newStatement.ifBlock
    }

    private fun lastElseStatement(): IfStatement {
        var ifStatement = this
        while (ifStatement.hasElseBranch) {
            if (ifStatement.elseStatement == null) {
                break // just to avoid infinity loop
            }

            ifStatement.elseStatement?.let {
                ifStatement = it
            }
        }
        return ifStatement
    }
}
